omp/components/atomic.hpp

```diff
@@ -32,10 +33,46 @@ void atomic_add(ValueType& out, ValueType val)
     // The C++ standard explicitly allows casting complex<double>* to double*
     // [complex.numbers.general]
     auto values = reinterpret_cast<gko::remove_complex<ValueType>*>(&out);
-#pragma omp atomic
-    values[0] += real(val);
-#pragma omp atomic
-    values[1] += imag(val);
+    atomic_add(values[0], real(val));
+    atomic_add(values[1], imag(val));
+}
+
+
+template <typename ResultType, typename ValueType>
+inline ResultType reinterpret(ValueType val)
+{
+    static_assert(sizeof(ValueType) == sizeof(ResultType),
+                  "The type to reinterpret to must be of the same size as the "
+                  "original type.");
+    return reinterpret_cast<ResultType&>(val);
```

Comment:
maybe just use `memcpy` here directly.

Reply:
I will change it but I do not think it help anything but we still need the same undefined behavior outside to make it work.